File: internal/providers/usage.go
package providers

// Token-usage extraction for the HTTP providers. Each vendor names the fields
// differently; this normalizes them onto types.TokenUsage.

import "github.com/overkazaf/re-agent/internal/types"

func anthropicUsage(raw map[string]any) types.TokenUsage {
	usage := asRecord(raw["usage"])
	if usage == nil {
		return types.TokenUsage{}
	}
	out := types.TokenUsage{}
	assign(&out.Input, usage["input_tokens"])
	assign(&out.Output, usage["output_tokens"])
	assign(&out.CacheRead, usage["cache_read_input_tokens"])
	assign(&out.CacheWrite, usage["cache_creation_input_tokens"])
	return out
}

func responsesUsage(raw map[string]any) types.TokenUsage {
	usage := asRecord(raw["usage"])
	if usage == nil {
		return types.TokenUsage{}
	}
	out := types.TokenUsage{}
	assign(&out.Input, usage["input_tokens"])
	assign(&out.Output, usage["output_tokens"])
	if details := asRecord(usage["output_tokens_details"]); details != nil {
		assign(&out.Thinking, details["reasoning_tokens"])
	}
	if cached := asRecord(usage["input_tokens_details"]); cached != nil {
		assign(&out.CacheRead, cached["cached_tokens"])
	}
	return out
}

func chatUsage(raw map[string]any) types.TokenUsage {
	usage := asRecord(raw["usage"])
	if usage == nil {
		return types.TokenUsage{}
	}
	out := types.TokenUsage{}
	assign(&out.Input, usage["prompt_tokens"])
	assign(&out.Output, usage["completion_tokens"])
	if details := asRecord(usage["completion_tokens_details"]); details != nil {
		assign(&out.Thinking, details["reasoning_tokens"])
	}
	if cached := asRecord(usage["prompt_tokens_details"]); cached != nil {
		assign(&out.CacheRead, cached["cached_tokens"])
	}
	return out
}

func assign(target *float64, value any) {
	if parsed, ok := num(value); ok {
		*target = parsed
	}
}

File: src/providers/usage.ts
// Token-usage extraction for the HTTP providers. Each vendor names the fields
// differently; this normalizes them onto TokenUsage.

import type { TokenUsage } from "../types";

function record(value: unknown): Record<string, unknown> | undefined {
  return value && typeof value === "object" ? (value as Record<string, unknown>) : undefined;
}

function num(value: unknown): number | undefined {
  return typeof value === "number" && Number.isFinite(value) ? value : undefined;
}

function compact(usage: TokenUsage): TokenUsage | undefined {
  return Object.values(usage).some(value => value !== undefined) ? usage : undefined;
}

export function anthropicUsage(raw: unknown): TokenUsage | undefined {
  const usage = record(record(raw)?.usage);
  if (!usage) return undefined;
  return compact({
    input: num(usage.input_tokens),
    output: num(usage.output_tokens),
    cacheRead: num(usage.cache_read_input_tokens),
    cacheWrite: num(usage.cache_creation_input_tokens),
  });
}

export function responsesUsage(raw: unknown): TokenUsage | undefined {
  const usage = record(record(raw)?.usage);
  if (!usage) return undefined;
  const details = record(usage.output_tokens_details);
  const cached = record(usage.input_tokens_details);
  return compact({
    input: num(usage.input_tokens),
    output: num(usage.output_tokens),
    thinking: num(details?.reasoning_tokens),
    cacheRead: num(cached?.cached_tokens),
  });
}

export function chatUsage(raw: unknown): TokenUsage | undefined {
  const usage = record(record(raw)?.usage);
  if (!usage) return undefined;
  const details = record(usage.completion_tokens_details);
  const cached = record(usage.prompt_tokens_details);
  return compact({
    input: num(usage.prompt_tokens),
    output: num(usage.completion_tokens),
    thinking: num(details?.reasoning_tokens),
    cacheRead: num(cached?.cached_tokens),
  });
}
